tanimoto_gp/__init__.py

```diff
@@ -63,3 +64,50 @@ def predict_y(self, params: TanimotoGP_Params, smiles_test: list[str], full_cova
         else:
             covar += TRANSFORM(params.raw_noise)
         return mean, covar
+
+    def _get_L(self, params: TanimotoGP_Params) -> jnp.ndarray:
+        """Get cached or compute new Cholesky factorization"""
+        if self._L_cached is None:
+            a = TRANSFORM(params.raw_amplitude)
+            s = TRANSFORM(params.raw_noise)
+            L = kgp._k_cholesky(
+                self._K_train_train,
+                s / a
+            )
+            self._L_cached = L
+        
+        return self._L_cached
+
+    def add_observation(self, params: TanimotoGP_Params, new_smiles: str, new_y: float) -> None:
+        """
+        Add a single observation and efficiently update cached Cholesky factorization.
+        Avoids recomputing K_train_train and Cholesky factorization repeatedly during BO.
+        """
+        new_fp = self._fp_func(new_smiles)
+
+        # Compute Tanimoto similarity between new observation and each existing training point
+        k_new = jnp.asarray(DataStructs.BulkTanimotoSimilarity(new_fp, self._fp_train))
+
+        # Diagonal term is Tanimoto simiarity of new observation w.r.t. itself
+        k_new_new = jnp.array(1.0)
+
+        # Update training data
+        self._smiles_train.append(new_smiles)
+        self._y_train = jnp.append(self._y_train, new_y)
+        self._fp_train.append(new_fp)
+
+        # Update kernel matrix
+        n = len(k_new)
+        top_block = jnp.concatenate([self._K_train_train, k_new.reshape(n, 1)], axis=1)
+        bottom_row = jnp.concatenate([k_new, k_new_new.reshape(1,)])
+        self._K_train_train = jnp.concatenate([top_block, bottom_row.reshape(1, n+1)], axis=0)
+
+        # If we have cached L, update it efficiently
+        if self._L_cached is not None:
+            a = TRANSFORM(params.raw_amplitude)
+            s = TRANSFORM(params.raw_noise)
+            self._L_cached = kgp.update_cholesky(
+                self._L_cached,
+                k_new,
+                k_new_new + (s/a)
+            )
```

Comment:
This looks good! However, would you consider generalizing it to adding `n` new points instead of just 1?